python/ray/data/dataset.py

```diff
@@ -461,6 +471,17 @@ def map_batches(
                 ``pandas.DataFrame``, "pyarrow" to select ``pyarrow.Table``, or
                 ``"numpy"`` to select ``numpy.ndarray`` for tensor datasets and
                 ``Dict[str, numpy.ndarray]`` for tabular datasets. Default is "default".
+            zero_copy_batch: Whether ``fn`` should be provided zero-copy, read-only
+                batches. If this is ``True`` and no copy is required for the
+                ``batch_format`` conversion, the batch will be a zero-copy, read-only
+                view on data in Ray's object store, which can decrease memory
```

Comment:
Orthogonal to this PR - if `map_batches` and `read` fused together, the block is in-memory and not a view from object store. Can we get rid of copy automatically when block is a view from object store?

Reply:
Yep I looked into this, among other optimizations!

The semantics are currently very conservative, where we tell the `Batcher` to ensure that batches are always fresh copies (if we're creating batches, i.e. `batch_size is not None`):
* https://github.com/ray-project/ray/blob/8f823cb51b8aee1e1fa8e351c70448d14bf771f5/python/ray/data/dataset.py#L570
* https://github.com/ray-project/ray/blob/8f823cb51b8aee1e1fa8e351c70448d14bf771f5/python/ray/data/_internal/batcher.py#L109-L112

This may create redundant copies if either

1. the base block is already in the worker heap (e.g. due to read and map_batches fusion, or due to the block not supporting zero-copy deserialization), or,
2. the batch conversion will already create a data copy.

The all-encompassing solution would be explicitly checking if the batch is read-only and only then making a copy, but for batch formats like Pandas DataFrames, this is a bit difficult/expensive: we'd need to do an `O(blocks) = O(columns x chunks)` traversal of the DataFrame to get the underlying (ndarray) blocks, and check if each block is writable. This appears to only be possible by using a lot of internal Pandas APIs, which will be super fragile:
```python
if all(block.array_values.to_numpy().flags["WRITEABLE"] for block in batch._data.blocks):
    # No copy needed.
else:
    # Copy needed.
```
This is less fragile if we converted the top-level columns to ndarrays and checked the `"WRITEABLE"` flag for those ndarrays, but this would often trigger block consolidation to create contiguous column ndarrays, resulting in a copy of the DataFrame just to check if we can get away with not copying the DataFrame, which is obviously a non-starter.

Reply:
For `zero_copy_batch=False`, I think that the following setup would be the best:
1. For batch formats that support relatively cheap (`O(columns)`) writability checks, such as NumPy and Arrow batches, only create a copy if the batch is not writable.
2. For other batch formats, i.e. Pandas and simple batches, elide copies if we are certain that the base block is already in the worker heap, or if we know that the block format --> batch format conversion will create a full copy.

Reply:
@clarkzinzow - got it, thanks. Maybe in optimizer, we can have a rule to auto-set `zero_copy_batch` if `map_batches` and `read` are fused together.

Reply:
@c21 I created an issue to capture this: https://github.com/ray-project/ray/issues/31073